sequencer/batchbuilder.go

```diff
@@ -497,26 +504,45 @@ func (s *Sequencer) appendPendingTxs(ctx context.Context, isClaims bool, minGasP
 			if !isClaims {
 				waitTick(ctx, ticker)
 			}
-			return 0
+			return nil
 		}
 	} else if err != nil {
 		log.Errorf("failed to get pending tx, err: %w", err)
-		return 0
+		return nil
 	}
-	var invalidTxsCounter int
+	var (
+		zkCountersBeforeAddition pool.ZkCounters
+		pendTxsPreSelected       []ethTypes.Transaction
+	)
 	for i := 0; i < len(pendTxs); i++ {
 		if pendTxs[i].FailedCounter > s.cfg.MaxAllowedFailedCounter {
 			hash := pendTxs[i].Transaction.Hash().String()
 			log.Warnf("mark tx with hash %s as invalid, failed counter %d exceeded max %d from config",
 				hash, pendTxs[i].FailedCounter, s.cfg.MaxAllowedFailedCounter)
 			s.updateTxsStatus(ctx, ticker, []string{hash}, pool.TxStatusInvalid)
-			invalidTxsCounter++
 			continue
 		}
-		s.sequenceInProgress.Txs = append(s.sequenceInProgress.Txs, pendTxs[i].Transaction)
+		if s.sequenceInProgress.ZkCounters.IsAnyFieldMoreThan(pool.ZkCounters{
+			CumulativeGasUsed:    s.cfg.MaxCumulativeGasUsed,
+			UsedKeccakHashes:     s.cfg.MaxKeccakHashes,
+			UsedPoseidonHashes:   s.cfg.MaxPoseidonHashes,
+			UsedPoseidonPaddings: s.cfg.MaxPoseidonPaddings,
+			UsedMemAligns:        s.cfg.MaxMemAligns,
+			UsedArithmetics:      s.cfg.MaxArithmetics,
+			UsedBinaries:         s.cfg.MaxBinaries,
+			UsedSteps:            s.cfg.MaxSteps,
+		}) {
+			log.Info("reached max zkCounters, need to close a batch")
+			s.sequenceInProgress.ZkCounters = zkCountersBeforeAddition
+			s.sequenceInProgress.IsZkCountersReachedMax = true
+			break
+		}
+		zkCountersBeforeAddition = s.sequenceInProgress.ZkCounters
+		s.sequenceInProgress.ZkCounters.SumUpZkCounters(pendTxs[i].ZkCounters)
+		pendTxsPreSelected = append(pendTxsPreSelected, pendTxs[i].Transaction)
```

Comment:
ZK counters should be used before executing (to estimate a realistic batch at first shot). Checking after executing should be redundant: if any counter is exceeded, executor shall return an OOC